ipam/allocator.go

```diff
@@ -159,389 +211,414 @@ const (
+func (a *Allocator) ReleasePool(poolID string) error {
+	k := SubnetKey{}
+	if err := k.FromString(poolID); err != nil {
+		return types.BadRequestErrorf("invalid pool id: %s", poolID)
+	}
 
-		if ipVer == v4 && i == num-1 {
-			// Do not let broadcast address be reserved
-			limit--
+retry:
+	remove, err := a.updatePoolDBOnRemoval(k)
+	if err != nil {
+		return err
+	}
+	if err = a.writeToStore(); err != nil {
+		if _, ok := err.(types.RetryError); !ok {
+			return types.InternalErrorf("pool (%s) removal failed because of %v", poolID, err)
 		}
-
-		// Generate the new address masks. AddressMask content may come from datastore
-		h, err := bitseq.NewHandle(dsDataKey, a.getStore(), smallKey.String(), limit)
-		if err != nil {
-			return err
+		if erru := a.readFromStore(); erru != nil {
+			return fmt.Errorf("failed to get updated pool config from datastore (%v) after (%v)", erru, err)
 		}
+		goto retry
+	}
 
-		if ipVer == v4 && i == 0 {
-			// Do not let network identifier address be reserved
-			h.Set(0)
-		}
+	return remove()
+}
 
-		a.Lock()
-		a.addresses[smallKey] = h
-		a.Unlock()
+func (a *Allocator) parsePoolRequest(addressSpace, pool, subPool string, v6 bool) (*SubnetKey, *net.IPNet, *net.IPNet, *AddressRange, error) {
+	var (
+		nw, aw *net.IPNet
+		ipr    *AddressRange
+		err    error
+	)
+
+	if addressSpace == "" {
+		return nil, nil, nil, nil, ipamapi.ErrInvalidAddressSpace
 	}
-	return nil
-}
 
-// Check subnets size. In case configured subnet is v6 and host size is
-// greater than 32 bits, adjust subnet to /96.
-func adjustAndCheckSubnetSize(subnet *net.IPNet) (*net.IPNet, error) {
-	ones, bits := subnet.Mask.Size()
-	if v6 == getAddressVersion(subnet.IP) {
-		if ones < minNetSizeV6 {
-			return nil, ErrInvalidSubnet
+	if pool == "" && subPool != "" {
+		return nil, nil, nil, nil, ipamapi.ErrInvalidSubPool
+	}
+
+	if pool != "" {
+		if _, nw, err = net.ParseCIDR(pool); err != nil {
+			return nil, nil, nil, nil, ipamapi.ErrInvalidPool
 		}
-		if ones < minNetSizeV6Eff {
-			newMask := net.CIDRMask(minNetSizeV6Eff, bits)
-			return &net.IPNet{IP: subnet.IP, Mask: newMask}, nil
+		if subPool != "" {
+			if ipr, err = getAddressRange(subPool); err != nil {
+				return nil, nil, nil, nil, err
+			}
 		}
 	} else {
-		if ones < minNetSize {
-			return nil, ErrInvalidSubnet
+		if nw, err = a.getPredefinedPool(addressSpace, v6); err != nil {
+			return nil, nil, nil, nil, err
 		}
+
 	}
-	return subnet, nil
+	if aw, err = adjustAndCheckSubnetSize(nw); err != nil {
+		return nil, nil, nil, nil, err
+	}
+
+	return &SubnetKey{AddressSpace: addressSpace, Subnet: nw.String(), ChildSubnet: subPool}, nw, aw, ipr, nil
 }
 
-// Checks whether the passed subnet is a superset or subset of any of the subset in the db
-func (a *Allocator) contains(space AddressSpace, subInfo *SubnetInfo) bool {
+func (a *Allocator) updatePoolDBOnAdd(k SubnetKey, ipr *AddressRange, nw *net.IPNet) (func() error, error) {
+	// Check if already allocated
```

Comment:
Let's keep this whole function under lock because there is a race when you are traversing the parent key tree while a `updatePoolDBonRemoval` is happening in parallel